viewing_party/party.py

```diff
@@ -1,23 +1,185 @@
+
+    return ratings_average
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
 
+# ---Wave_2_function_2---
+def get_most_watched_genre(user_data):
+
+    movie_genre_dict = {}
+
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    for movie in user_data["watched"]:
+
+        
+        title = movie["title"]
+        genre = movie["genre"]
+        
+        if genre not in movie_genre_dict:
+            movie_genre_dict[genre] = []
         
-# -----------------------------------------
+        movie_genre_dict[genre].append(title)
+
+    max_genre = None
+    max_count = 0
+
+    for genre in movie_genre_dict:
+        genre_length = len(movie_genre_dict[genre])
+        if genre_length >= max_count:
+            max_genre = genre
+            max_count = len(movie_genre_dict.values())
+
+    return max_genre
+
+# ------------- WAVE 3 --------------------
+
+#---Wave_3_function_1---
+
+def get_unique_watched(user_data):
+    user_unique_movies = []
+    movies_friends_watched = []
+    movies_user_watched = user_data["watched"]
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            movies_friends_watched.append(movie)
+    
+    for movie in movies_user_watched:
+        if movie not in movies_friends_watched:
+            user_unique_movies.append(movie)
+
+    return user_unique_movies
+
+#---Wave_3_function_2---
+
+def get_friends_unique_watched(user_data):
+
+    movies_user_watched = user_data["watched"]
+    unique_movies_friends_watched = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if (movie not in unique_movies_friends_watched) and (movie not in movies_user_watched):
+                unique_movies_friends_watched.append(movie)
+    
+    # returns list of movie dictionaries
+    return unique_movies_friends_watched
+
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
+#---Wave_4_function_1---
+
+def get_available_recs(user_data):
+
+    friends_unique_movie_list = get_friends_unique_watched(user_data)
+
+    # figure out if host is in user's subscriptions
+    recommended_movies = []
+    for movie in friends_unique_movie_list:
+        if movie["host"] in user_data["subscriptions"]:
+            recommended_movies.append(movie)
+    
+    return recommended_movies
 
-# -----------------------------------------
 # ------------- WAVE 5 --------------------
-# -----------------------------------------
+# ---Wave_5_function_1---
+def get_new_rec_by_genre(user_data):
+    rec_by_genre = []
+
+    users_watched = user_data["watched"]
+    friends_watched = get_friends_unique_watched(user_data)
+    users_most_freq_genre = get_most_watched_genre(user_data)
 
+    if len(users_watched) == 0:
+        return rec_by_genre
+
+    if len(friends_watched) == 0:
+        return rec_by_genre

```

Comment:
Actually, we can get rid of this conditional and the tests still pass! If `friends_watched` is empty, then the code skips the below `for` loop, but still returns `rec_by_genre` (which is still an `[]`)